Source/Android/app/src/main/java/org/dolphinemu/dolphinemu/model/GameFileCache.java

```diff
@@ -28,67 +22,89 @@ public GameFileCache(String path)
+  private static LinkedHashSet<String> getPathSet(boolean removeNonExistentFolders)
   {
-    SharedPreferences preferences = PreferenceManager.getDefaultSharedPreferences(context);
-    Set<String> folderPaths = preferences.getStringSet(GAME_FOLDER_PATHS_PREFERENCE, EMPTY_SET);
+    File dolphinFile = SettingsFile.getSettingsFile(Settings.FILE_DOLPHIN);
+    IniFile dolphinIni = new IniFile(dolphinFile);
+    LinkedHashSet<String> pathSet = new LinkedHashSet<>();
+    int totalISOPaths =
+            dolphinIni.getInt(Settings.SECTION_INI_GENERAL, SettingsFile.KEY_ISO_PATHS, 0);
 
-    if (folderPaths == null)
+    if (totalISOPaths < 1)
     {
-      return;
+      return pathSet;
     }
 
-    Set<String> newFolderPaths = new HashSet<>();
-    for (String folderPath : folderPaths)
+    for (int i = 0; i < totalISOPaths; i++)
     {
-      File folder = new File(folderPath);
+      String path = dolphinIni.getString(Settings.SECTION_INI_GENERAL,
+              SettingsFile.KEY_ISO_PATH_BASE + i, "");
+
+      File folder = new File(path);
       if (folder.exists())
       {
-        newFolderPaths.add(folderPath);
+        pathSet.add(path);
       }
     }
 
-    if (folderPaths.size() != newFolderPaths.size())
+    if (removeNonExistentFolders && totalISOPaths > pathSet.size())
     {
-      // One or more folders are being deleted
-      SharedPreferences.Editor editor = preferences.edit();
-      editor.putStringSet(GAME_FOLDER_PATHS_PREFERENCE, newFolderPaths);
-      editor.apply();
+      int setIndex = 0;
+
+      dolphinIni.setInt(Settings.SECTION_INI_GENERAL, SettingsFile.KEY_ISO_PATHS,
+              pathSet.size());
+
+      // One or more folders have been removed.
+      for (String entry : pathSet)
+      {
+        dolphinIni.setString(Settings.SECTION_INI_GENERAL, SettingsFile.KEY_ISO_PATH_BASE +
+                setIndex, entry);
+
+        setIndex = setIndex + 1;
+      }
+
+      // Delete known unnecessary keys. Ignore i values beyond index-adjusted totalISOPaths.
+      for (int i = totalISOPaths - pathSet.size() - 1; i < totalISOPaths; i++)
+      {
+        dolphinIni.deleteKey(Settings.SECTION_INI_GENERAL, SettingsFile.KEY_ISO_PATH_BASE + i);
+      }
+
+      dolphinIni.save(dolphinFile);
     }
+
+    return pathSet;
   }
 
   /**
    * Scans through the file system and updates the cache to match.
    *
    * @return true if the cache was modified
    */
-  public boolean scanLibrary(Context context)
+  public boolean scanLibrary()
   {
     boolean recursiveScan = BooleanSetting.MAIN_RECURSIVE_ISO_PATHS.getBooleanGlobal();
 
-    removeNonExistentGameFolders(context);
-
-    SharedPreferences preferences = PreferenceManager.getDefaultSharedPreferences(context);
-    Set<String> folderPathsSet = preferences.getStringSet(GAME_FOLDER_PATHS_PREFERENCE, EMPTY_SET);
+    LinkedHashSet<String> folderPathsSet = getPathSet(true);
 
-    if (folderPathsSet == null)
+    if (folderPathsSet.isEmpty())
```

Comment:
This check can be removed, and I believe it in fact has to be removed in order to correctly handle the case where all folders have gotten deleted from the paths set.

Reply:
I guess the original intention of that check was to skip updating the cache if that update would not show games. I agree that is something we'd want to change.

Reply:
I think the original intention was just to avoid a null pointer dereference. Note that it only checked for the null case and not for the empty case.

Reply:
Oh right, I forgot I changed the initialization of the set.